containers/src/Data/Set/Internal.hs

```diff
@@ -1855,8 +1862,25 @@ cartesianProduct :: Set a -> Set b -> Set (a, b)
 -- Unfortunately, this is much slower in practice, at least when the sets are
 -- constructed from ascending lists. I tried doing the same thing using a
 -- known-length (perfect balancing) variant of fromDistinctAscList, but it
--- still didn't come close to the performance of Kmett's version in my very
--- informal tests.
+-- still didn't come close to the performance of the implementation we use in my
+-- very informal tests.
+--
+-- The implementation we use (slightly modified from one that Edward Kmett
+-- hacked together) is also optimal but performs better in practice. We map
+-- each element a in as to a set made up of (a,b) for every element b in bs,
+-- taking O(nm) overall. Then we merge these sets up the tree of as, which takes
+-- O(n log m). A brief sketch of proof for the latter:
+--
+-- There are O(n/2^i) nodes at level i of the tree. At a node at level i, we
+-- merge the merged left set, current set, and merged right set into a set of
+-- size O(2^i*m). This is done in a logarithmic time of O(i + log m). Over all
+-- levels, we do a total work of
+--
+--    O(\sum_{i=1}^{height} n * (i + log m) / 2^i)
+--  = O(  n * \sum_{i=1}^{inf} i/2^i
+--      + n * log m * \sum_{i=1}^{inf} 1/2^i)
```

Comment:
Could you add an extra step? Switching the limits of summation and distributing are unrelated.

Reply:
Done.